File: enchiridion-go/internal/watch/watch_test.go
package watch

import (
	"os"
	"path/filepath"
	"reflect"
	"sync"
	"testing"
	"time"
)

// --- debounce timing ---------------------------------------------------------

func TestDebounceNotSettledWithinWindow(t *testing.T) {
	now := 0.0
	d := NewDebouncer(30.0, func() float64 { return now })

	for _, ts := range []float64{0.0, 5.0, 10.0, 15.0, 20.0, 25.0} {
		now = ts
		d.RecordEvent("raw/notes/a.md")
	}
	if got := d.SettledFiles(); len(got) != 0 {
		t.Errorf("settled = %v, want empty", got)
	}
}

func TestDebounceSettlesAfterFinalSilence(t *testing.T) {
	now := 0.0
	d := NewDebouncer(30.0, func() float64 { return now })

	for _, ts := range []float64{0.0, 10.0, 20.0, 35.0} {
		now = ts
		d.RecordEvent("raw/notes/a.md")
	}
	now = 35.0 + 29.0
	if got := d.SettledFiles(); len(got) != 0 {
		t.Errorf("settled early = %v", got)
	}
	now = 35.0 + 30.0
	if got := d.SettledFiles(); !reflect.DeepEqual(got, []string{"raw/notes/a.md"}) {
		t.Errorf("settled = %v", got)
	}
}

func TestDebounceSettledFilesStopBeingTracked(t *testing.T) {
	now := 0.0
	d := NewDebouncer(10.0, func() float64 { return now })
	d.RecordEvent("raw/a.md")
	now = 10.0
	if got := d.SettledFiles(); !reflect.DeepEqual(got, []string{"raw/a.md"}) {
		t.Errorf("first = %v", got)
	}
	now = 100.0
	if got := d.SettledFiles(); len(got) != 0 {
		t.Errorf("re-reported without a new event: %v", got)
	}
}

func TestDebounceIsPerFile(t *testing.T) {
	now := 0.0
	d := NewDebouncer(30.0, func() float64 { return now })
	d.RecordEvent("raw/a.md")
	now = 15.0
	d.RecordEvent("raw/b.md")
	now = 30.0
	if got := d.SettledFiles(); !reflect.DeepEqual(got, []string{"raw/a.md"}) {
		t.Errorf("settled = %v, want only a.md", got)
	}
}

// --- lock file lifecycle -----------------------------------------------------

func TestWriteLockThenRemove(t *testing.T) {
	lockPath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch.lock")
	if err := WriteLock(lockPath, 1234, time.Time{}); err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(lockPath); err != nil {
		t.Fatalf("lock not written: %v", err)
	}
	if err := RemoveLock(lockPath); err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(lockPath); !os.IsNotExist(err) {
		t.Error("lock not removed")
	}
}

func TestRemoveLockMissingIsANoop(t *testing.T) {
	if err := RemoveLock(filepath.Join(t.TempDir(), "watch.lock")); err != nil {
		t.Fatal(err)
	}
}

func TestAcquireLockLivePIDBails(t *testing.T) {
	lockPath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch.lock")
	if err := WriteLock(lockPath, os.Getpid(), time.Time{}); err != nil {
		t.Fatal(err)
	}
	ok, _, err := AcquireLock(lockPath, time.Time{}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("live lock reported acquired")
	}
}

func TestAcquireLockOldTimestampIsStale(t *testing.T) {
	lockPath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch.lock")
	old := time.Now().UTC().Add(-11 * time.Minute)
	if err := WriteLock(lockPath, os.Getpid(), old); err != nil {
		t.Fatal(err)
	}
	ok, _, err := AcquireLock(lockPath, time.Time{}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("stale (old timestamp) lock not acquired")
	}
}

func TestAcquireLockRecentTimestampNotStale(t *testing.T) {
	lockPath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch.lock")
	recent := time.Now().UTC().Add(-9 * time.Minute)
	if err := WriteLock(lockPath, os.Getpid(), recent); err != nil {
		t.Fatal(err)
	}
	ok, _, err := AcquireLock(lockPath, time.Time{}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("recent live lock reported acquired")
	}
}

func TestAcquireLockNoExistingLockSucceeds(t *testing.T) {
	lockPath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch.lock")
	ok, _, err := AcquireLock(lockPath, time.Time{}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("empty vault did not acquire")
	}
}

func TestAcquireLockUnparsableLockIsStale(t *testing.T) {
	lockPath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch.lock")
	if err := os.MkdirAll(filepath.Dir(lockPath), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(lockPath, []byte("not json"), 0o644); err != nil {
		t.Fatal(err)
	}
	ok, _, err := AcquireLock(lockPath, time.Time{}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("unparsable lock not acquired")
	}
}

func TestAcquireLockDeadPIDIsStale(t *testing.T) {
	lockPath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch.lock")
	// A pid that is (almost certainly) not alive on any test host.
	if err := WriteLock(lockPath, 1<<30, time.Time{}); err != nil {
		t.Fatal(err)
	}
	ok, stalePID, err := AcquireLock(lockPath, time.Time{}, nil)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("dead-pid lock not acquired")
	}
	if stalePID == nil || *stalePID != 1<<30 {
		t.Errorf("stalePID = %v, want the removed lock's pid", stalePID)
	}
}

func TestAcquireLockConcurrentStaleTakeoverOnlyOneWinner(t *testing.T) {
	lockPath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch.lock")
	if err := WriteLock(lockPath, 1<<30, time.Time{}); err != nil {
		t.Fatal(err)
	}

	const n = 8
	start := make(chan struct{})
	var mu sync.Mutex
	var winners []bool
	var wg sync.WaitGroup
	for i := 0; i < n; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			<-start
			ok, _, err := AcquireLock(lockPath, time.Time{}, nil)
			if err != nil {
				t.Errorf("AcquireLock: %v", err)
				return
			}
			mu.Lock()
			winners = append(winners, ok)
			mu.Unlock()
		}()
	}
	close(start)
	wg.Wait()

	count := 0
	for _, w := range winners {
		if w {
			count++
		}
	}
	if count != 1 {
		t.Errorf("%d winners, want exactly 1", count)
	}
}

// --- queue -------------------------------------------------------------------

func TestQueueAppendCreatesAndAppends(t *testing.T) {
	queuePath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch-queue.jsonl")
	if err := AppendQueue(queuePath, "raw/a.md"); err != nil {
		t.Fatal(err)
	}
	if err := AppendQueue(queuePath, "raw/b.md"); err != nil {
		t.Fatal(err)
	}
	got, err := ReadQueue(queuePath)
	if err != nil {
		t.Fatal(err)
	}
	if want := []string{"raw/a.md", "raw/b.md"}; !reflect.DeepEqual(got, want) {
		t.Errorf("queue = %v, want %v", got, want)
	}
}

func TestQueueAppendIsIdempotent(t *testing.T) {
	queuePath := filepath.Join(t.TempDir(), "watch-queue.jsonl")
	if err := AppendQueue(queuePath, "raw/a.md"); err != nil {
		t.Fatal(err)
	}
	if err := AppendQueue(queuePath, "raw/a.md"); err != nil {
		t.Fatal(err)
	}
	got, _ := ReadQueue(queuePath)
	if want := []string{"raw/a.md"}; !reflect.DeepEqual(got, want) {
		t.Errorf("queue = %v, want %v", got, want)
	}
}

func TestQueueRemove(t *testing.T) {
	queuePath := filepath.Join(t.TempDir(), "watch-queue.jsonl")
	for _, rel := range []string{"raw/a.md", "raw/b.md"} {
		if err := AppendQueue(queuePath, rel); err != nil {
			t.Fatal(err)
		}
	}
	if err := RemoveFromQueue(queuePath, "raw/a.md"); err != nil {
		t.Fatal(err)
	}
	got, _ := ReadQueue(queuePath)
	if want := []string{"raw/b.md"}; !reflect.DeepEqual(got, want) {
		t.Errorf("queue = %v, want %v", got, want)
	}
}

func TestQueueReadMissingFileIsEmpty(t *testing.T) {
	got, err := ReadQueue(filepath.Join(t.TempDir(), "watch-queue.jsonl"))
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 0 {
		t.Errorf("queue = %v, want empty", got)
	}
}

func TestQueueConcurrentAppendsDoNotCorrupt(t *testing.T) {
	queuePath := filepath.Join(t.TempDir(), ".wiki-knowledge", "watch-queue.jsonl")
	rels := []string{"raw/a.md", "raw/b.md", "raw/c.md", "raw/d.md", "raw/e.md", "raw/f.md"}
	var wg sync.WaitGroup
	for _, rel := range rels {
		wg.Add(1)
		go func(rel string) {
			defer wg.Done()
			_ = AppendQueue(queuePath, rel)
		}(rel)
	}
	wg.Wait()

	got, _ := ReadQueue(queuePath)
	if len(got) != len(rels) {
		t.Errorf("queue has %d entries, want %d (lost or mangled): %v", len(got), len(rels), got)
	}
	seen := map[string]bool{}
	for _, line := range got {
		if seen[line] {
			t.Errorf("duplicate line %q", line)
		}
		seen[line] = true
	}
}

// --- eligibility check on settle ----------------------------------------------

func TestCheckAndEnqueueEnqueuesWhenEligible(t *testing.T) {
	queuePath := filepath.Join(t.TempDir(), "watch-queue.jsonl")
	ok, err := CheckAndEnqueue(map[string]bool{"raw/a.md": true}, "raw/a.md", queuePath)
	if err != nil || !ok {
		t.Fatalf("ok=%v err=%v", ok, err)
	}
	got, _ := ReadQueue(queuePath)
	if want := []string{"raw/a.md"}; !reflect.DeepEqual(got, want) {
		t.Errorf("queue = %v", got)
	}
}

func TestCheckAndEnqueueSkipsWhenNotEligible(t *testing.T) {
	queuePath := filepath.Join(t.TempDir(), "watch-queue.jsonl")
	ok, err := CheckAndEnqueue(map[string]bool{"raw/other.md": true}, "raw/a.md", queuePath)
	if err != nil || ok {
		t.Fatalf("ok=%v err=%v", ok, err)
	}
	got, _ := ReadQueue(queuePath)
	if len(got) != 0 {
		t.Errorf("queue = %v, want empty", got)
	}
}

// --- RelForEvent --------------------------------------------------------------

func TestRelForEventMapsFileUnderRoot(t *testing.T) {
	root := t.TempDir()
	path := filepath.Join(root, "raw", "note.md")
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(path, []byte("x"), 0o644); err != nil {
		t.Fatal(err)
	}
	rel, ok := RelForEvent(root, path)
	if !ok || rel != "raw/note.md" {
		t.Errorf("RelForEvent = (%q, %v)", rel, ok)
	}
}

func TestRelForEventIgnoresDirectory(t *testing.T) {
	root := t.TempDir()
	dir := filepath.Join(root, "raw", "subdir")
	if err := os.MkdirAll(dir, 0o755); err != nil {
		t.Fatal(err)
	}
	if _, ok := RelForEvent(root, dir); ok {
		t.Error("directory reported as a file event")
	}
}

func TestRelForEventIgnoresPathOutsideRoot(t *testing.T) {
	root := t.TempDir()
	outside := filepath.Join(t.TempDir(), "note.md")
	if err := os.WriteFile(outside, []byte("x"), 0o644); err != nil {
		t.Fatal(err)
	}
	if _, ok := RelForEvent(root, outside); ok {
		t.Error("outside-root path reported as a file event")
	}
}

File: enchiridion-ts/src/watch.test.ts
/**
 * watch tests — the debounce/lock/queue machinery behind /wiki-watch.
 */

import { test } from "node:test";
import assert from "node:assert/strict";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import {
  DefaultPollIntervalSeconds,
  DefaultDebounceSeconds,
  Debouncer,
  StaleLockSeconds,
  appendQueue,
  acquireLock,
  checkAndEnqueue,
  forRoot,
  readQueue,
  relForEvent,
  removeFromQueue,
  removeLock,
  runWatch,
  writeLock,
  type Watcher,
} from "./watch.js";

function tmpRoot(): string {
  return fs.mkdtempSync(path.join(os.tmpdir(), "watch-test-"));
}

// --- debounce timing ---------------------------------------------------------

test("debounce: not settled within window", () => {
  let now = 0;
  const d = new Debouncer(30, () => now);
  for (const ts of [0, 5, 10, 15, 20, 25]) {
    now = ts;
    d.recordEvent("raw/notes/a.md");
  }
  assert.deepEqual(d.settledFiles(), []);
});

test("debounce: settles after final silence", () => {
  let now = 0;
  const d = new Debouncer(30, () => now);
  for (const ts of [0, 10, 20, 35]) {
    now = ts;
    d.recordEvent("raw/notes/a.md");
  }
  now = 35 + 29;
  assert.deepEqual(d.settledFiles(), []);
  now = 35 + 30;
  assert.deepEqual(d.settledFiles(), ["raw/notes/a.md"]);
});

test("debounce: settled files stop being tracked", () => {
  let now = 0;
  const d = new Debouncer(10, () => now);
  d.recordEvent("raw/a.md");
  now = 10;
  assert.deepEqual(d.settledFiles(), ["raw/a.md"]);
  now = 100;
  assert.deepEqual(d.settledFiles(), []);
});

test("debounce: is per-file", () => {
  let now = 0;
  const d = new Debouncer(30, () => now);
  d.recordEvent("raw/a.md");
  now = 15;
  d.recordEvent("raw/b.md");
  now = 30;
  assert.deepEqual(d.settledFiles(), ["raw/a.md"]);
});

// --- lock file lifecycle -----------------------------------------------------

test("write lock then remove", () => {
  const lockPath = path.join(tmpRoot(), ".wiki-knowledge", "watch.lock");
  writeLock(lockPath, 1234, new Date());
  assert.ok(fs.existsSync(lockPath));
  removeLock(lockPath);
  assert.ok(!fs.existsSync(lockPath));
});

test("remove lock missing is a no-op", () => {
  assert.doesNotThrow(() => removeLock(path.join(tmpRoot(), "watch.lock")));
});

test("acquire lock: live PID bails", () => {
  const lockPath = path.join(tmpRoot(), ".wiki-knowledge", "watch.lock");
  writeLock(lockPath, process.pid, new Date());
  const { acquired } = acquireLock(lockPath);
  assert.equal(acquired, false);
});

test("acquire lock: old timestamp is stale", () => {
  const lockPath = path.join(tmpRoot(), ".wiki-knowledge", "watch.lock");
  const old = new Date(Date.now() - (StaleLockSeconds + 60) * 1000);
  writeLock(lockPath, process.pid, old);
  const { acquired } = acquireLock(lockPath);
  assert.equal(acquired, true);
});

test("acquire lock: recent timestamp not stale", () => {
  const lockPath = path.join(tmpRoot(), ".wiki-knowledge", "watch.lock");
  const recent = new Date(Date.now() - (StaleLockSeconds - 60) * 1000);
  writeLock(lockPath, process.pid, recent);
  const { acquired } = acquireLock(lockPath);
  assert.equal(acquired, false);
});

test("acquire lock: no existing lock succeeds", () => {
  const lockPath = path.join(tmpRoot(), ".wiki-knowledge", "watch.lock");
  const { acquired, stalePID } = acquireLock(lockPath);
  assert.equal(acquired, true);
  assert.equal(stalePID, null);
});

test("acquire lock: unparsable lock is stale", () => {
  const lockPath = path.join(tmpRoot(), ".wiki-knowledge", "watch.lock");
  fs.mkdirSync(path.dirname(lockPath), { recursive: true });
  fs.writeFileSync(lockPath, "not json");
  const { acquired } = acquireLock(lockPath);
  assert.equal(acquired, true);
});

test("acquire lock: dead PID is stale, reports the removed pid", () => {
  const lockPath = path.join(tmpRoot(), ".wiki-knowledge", "watch.lock");
  const deadPID = 1 << 30;
  writeLock(lockPath, deadPID, new Date());
  const { acquired, stalePID } = acquireLock(lockPath);
  assert.equal(acquired, true);
  assert.equal(stalePID, deadPID);
});

// --- queue -------------------------------------------------------------------

test("queue: append creates and appends", () => {
  const queuePath = path.join(
    tmpRoot(),
    ".wiki-knowledge",
    "watch-queue.jsonl",
  );
  appendQueue(queuePath, "raw/a.md");
  appendQueue(queuePath, "raw/b.md");
  assert.deepEqual(readQueue(queuePath), ["raw/a.md", "raw/b.md"]);
});

test("queue: append is idempotent", () => {
  const queuePath = path.join(tmpRoot(), "watch-queue.jsonl");
  appendQueue(queuePath, "raw/a.md");
  appendQueue(queuePath, "raw/a.md");
  assert.deepEqual(readQueue(queuePath), ["raw/a.md"]);
});

test("queue: remove", () => {
  const queuePath = path.join(tmpRoot(), "watch-queue.jsonl");
  appendQueue(queuePath, "raw/a.md");
  appendQueue(queuePath, "raw/b.md");
  removeFromQueue(queuePath, "raw/a.md");
  assert.deepEqual(readQueue(queuePath), ["raw/b.md"]);
});

test("queue: read missing file is empty", () => {
  assert.deepEqual(readQueue(path.join(tmpRoot(), "watch-queue.jsonl")), []);
});

// --- eligibility check on settle ----------------------------------------------

test("check-and-enqueue: enqueues when eligible", () => {
  const queuePath = path.join(tmpRoot(), "watch-queue.jsonl");
  const ok = checkAndEnqueue(new Set(["raw/a.md"]), "raw/a.md", queuePath);
  assert.equal(ok, true);
  assert.deepEqual(readQueue(queuePath), ["raw/a.md"]);
});

test("check-and-enqueue: skips when not eligible", () => {
  const queuePath = path.join(tmpRoot(), "watch-queue.jsonl");
  const ok = checkAndEnqueue(new Set(["raw/other.md"]), "raw/a.md", queuePath);
  assert.equal(ok, false);
  assert.deepEqual(readQueue(queuePath), []);
});

// --- relForEvent --------------------------------------------------------------

test("rel-for-event: maps file under root", () => {
  const root = tmpRoot();
  const p = path.join(root, "raw", "note.md");
  fs.mkdirSync(path.dirname(p), { recursive: true });
  fs.writeFileSync(p, "x");
  assert.equal(relForEvent(root, p), "raw/note.md");
});

test("rel-for-event: ignores directory", () => {
  const root = tmpRoot();
  const dir = path.join(root, "raw", "subdir");
  fs.mkdirSync(dir, { recursive: true });
  assert.equal(relForEvent(root, dir), null);
});

test("rel-for-event: ignores path outside root", () => {
  const root = tmpRoot();
  const outside = path.join(tmpRoot(), "note.md");
  fs.writeFileSync(outside, "x");
  assert.equal(relForEvent(root, outside), null);
});

// --- paths + defaults ----------------------------------------------------------

test("for-root: yields the watch paths under .wiki-knowledge", () => {
  const paths = forRoot("/vault");
  assert.equal(paths.lock, "/vault/.wiki-knowledge/watch.lock");
  assert.equal(paths.queue, "/vault/.wiki-knowledge/watch-queue.jsonl");
});

test("defaults: exported constants", () => {
  assert.equal(DefaultDebounceSeconds, 30);
  assert.equal(DefaultPollIntervalSeconds, 5);
  assert.equal(StaleLockSeconds, 600);
});

// --- the watch loop (runWatch seam tests) ------------------------------------

/** A fake file watcher: the test fires "ready"/"all" events and calls
 * close(), with no chokidar on the loop's seam. */
class FakeWatcher implements Watcher {
  closed = false;
  private readyHandlers: (() => void)[] = [];
  private allHandlers: ((eventName: string, p: string) => void)[] = [];
  private errorHandlers: ((err: unknown) => void)[] = [];
  on(
    event: "all" | "error" | "ready",
    cb: (eventName: string, p: string) => void,
  ): void {
    if (event === "all") this.allHandlers.push(cb);
    else if (event === "error")
      this.errorHandlers.push(cb as (err: unknown) => void);
    else this.readyHandlers.push(cb as () => void);
  }
  close(): Promise<void> {
    this.closed = true;
    return Promise.resolve();
  }
  emit(event: string, ...args: unknown[]): void {
    if (event === "all") {
      for (const cb of this.allHandlers)
        cb(args[0] as string, args[1] as string);
    } else if (event === "error") {
      for (const cb of this.errorHandlers) cb(args[0]);
    } else {
      for (const cb of this.readyHandlers) cb();
    }
  }
}

/** Await a macrotask so the tick's async sweep chain has run to completion. */
async function flush(): Promise<void> {
  await new Promise<void>((r) => setImmediate(r));
}

/** A fake signal hub: runWatch's onSignal/offSignal pair, so a test can fire
 * SIGINT/SIGTERM without touching the process. */
function recordSignals(): {
  onSignal: (sig: "SIGINT" | "SIGTERM", cb: () => void) => void;
  offSignal: (sig: "SIGINT" | "SIGTERM", cb: () => void) => void;
  signals: Record<string, (() => void)[]>;
} {
  const signals: Record<string, (() => void)[]> = {};
  return {
    onSignal: (sig, cb) => {
      (signals[sig] ??= []).push(cb);
    },
    offSignal: (sig, cb) => {
      signals[sig] = (signals[sig] ?? []).filter((c) => c !== cb);
    },
    signals,
  };
}

/** Writes a real file under root/raw/, returning its absolute path — the loop
 * maps it through relForEvent exactly as production would. */
function rawFile(root: string, name: string): string {
  const abs = path.join(root, "raw", name);
  fs.mkdirSync(path.dirname(abs), { recursive: true });
  fs.writeFileSync(abs, "x");
  return abs;
}

test("run-watch: settle, sweep, enqueue end to end; a signal stops and cleans up", async () => {
  const root = tmpRoot();
  const paths = forRoot(root);
  const abs = rawFile(root, "a.md");

  const watcher = new FakeWatcher();
  const { onSignal, offSignal, signals } = recordSignals();
  const lines: string[] = [];
  const ticks: (() => void)[] = [];
  let pollMs = 0;
  let now = 0;
  let sweepRuns = 0;

  const done = runWatch(paths, {
    debounceSeconds: 30,
    pollIntervalSeconds: 5,
    makeWatcher: () => watcher,
    clock: () => now,
    sweep: async () => {
      sweepRuns++;
      return new Set(["raw/a.md"]);
    },
    onSignal,
    offSignal,
    schedule: (cb, ms) => {
      ticks.push(cb);
      pollMs = ms;
      return () => {};
    },
    pid: 42,
    log: (l) => lines.push(l),
  });

  assert.equal(pollMs, 5 * 1000);
  watcher.emit("ready");
  watcher.emit("all", "add", abs);
  now = 100;
  ticks[0]();
  await flush();

  assert.deepEqual(readQueue(paths.queue), ["raw/a.md"]);
  assert.equal(sweepRuns, 1);
  assert.deepEqual(lines, [
    `watching ${path.join(root, "raw")} (debounce=30s, pid=42)`,
    "queued raw/a.md",
  ]);

  writeLock(paths.lock, 999, new Date());
  signals.SIGTERM?.[0]();
  await done;
  assert.match(lines[lines.length - 1], /watcher stopped/);
  assert.ok(!fs.existsSync(paths.lock), "lock removed on stop");
  assert.equal(watcher.closed, true);
  assert.deepEqual(signals.SIGINT, []);
  assert.deepEqual(signals.SIGTERM, []);
});

test("run-watch: an eligibility miss settles without enqueuing", async () => {
  const root = tmpRoot();
  const paths = forRoot(root);
  const abs = rawFile(root, "b.md");

  const watcher = new FakeWatcher();
  const { onSignal, offSignal, signals } = recordSignals();
  const ticks: (() => void)[] = [];
  let now = 0;
  const sweepSets: Set<string>[] = [
    new Set(["raw/other.md"]),
    new Set(["raw/b.md"]),
  ];
  let sweepRuns = 0;

  const done = runWatch(paths, {
    debounceSeconds: 30,
    makeWatcher: () => watcher,
    clock: () => now,
    sweep: async () => sweepSets[sweepRuns++],
    onSignal,
    offSignal,
    schedule: (cb) => {
      ticks.push(cb);
      return () => {};
    },
    log: () => {},
  });

  watcher.emit("all", "add", abs);
  now = 100;
  ticks[0]();
  await flush();
  assert.deepEqual(readQueue(paths.queue), []);
  assert.equal(sweepRuns, 1);

  // A settled-and-missed file stops being tracked: the next tick has nothing
  // to sweep, even though the sweep would now offer the file.
  ticks[0]();
  await flush();
  assert.equal(sweepRuns, 1);
  assert.deepEqual(readQueue(paths.queue), []);

  signals.SIGTERM?.[0]();
  await done;
});
